fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+for these expressions.
+
+## High Level Operation
+
+Each N to M tuple mapping is expressed in a named FSL transform specification.
+
+The input to a transform is a sequence of instructions. The output of 
+a transform is a possibly transformed sequence. The input sequence can
+be modified in place or into a separate buffer. This is controlled by 
+FSL syntax.
+
+The operation of an FSL transform consists of a matching phase,
+a constraints checking phase, and the conversion phase.  The transform 
+phases are represented by three clauses in an FSL transform specification. 
+
+A phase returns an indication of success or failure. If a phase reports
+a failure, control is returned to the Fusion API. The API will continue
+processing and notify the system as required.
+
+The sequence clause declares the matching attributes to be applied to 
+the input sequence. The matching attributes are a list of
+instruction objects, either as assembly language or as a list of UIDs.
+
+The output of the sequence clause indicates which instructions
+are to be transformed, if any. If no match is found the sequence clause
+returns fail.
+
+In the simplest case, a positive match is indicated by an index and a length.
+The index is the position of the first matching instruction in the input
+sequence, length is the number of instructions to be transformed, inclusive of
+the index.
+
+A valid sequence indication is passed to the constraints clause. Subsequently,
+the constraints specification is applied to the indicated range and a 
+true/false flag is returned.
+
+If the constraints are satisfied, the constraints clause passes true to the
+conversion clause, along with the sequence indication.
+
+If the conversion clause receives true, it will apply the transformation
+on the input sequence beginning at the specified index and places the result 
+in an output sequence or modifies the input sequence in place as directed
+by syntax.
+
+## FSL Example
+
+An example of an FSL transform specification written in the encapsulated
+style is shown below.
+
+The three clauses, sequence, constraints and conversion, are preceded by
+the variables which declare the instruction set architecture, 
+the micro-architecture of the processor and a 
+sequence container reference, ioput. ioput links the transform to the
+performance model's decoded instructions containers.
+
+The isa/uarch/ioput are referred to as the 'Prolog' in discussions
+below.
+
+```
+transform fs1 {
+
+  isa   rv64g
+  uarch oly1
+  ioput iop1
+
+  sequence seq1(iop1,rv64g) {
+    c.lui    g1, c1
+    c.addi   g1, g1, c2
+    _req_ 2
+    c.xor    g2, g2, g1
+    c.slli   g2, g2, c3
+    _opt_ 1
+    c.srli   g2,     c3
+  }
+
+  constraints cons1(iop1,seq1,rv64g,oly1) {
+    gpr g1,g2
+    s20 c1
+    s12 c2
+    u6  c3
+
+    g1 != g2
+  }
+
+  conversion conv1(iop1,seq1,cons1) {
+    instr fused(opcode=0x1234)
+    iop1.input.replace(seq1,fused)
+  }
+}
+```
```

Comment:
If we do not anticipate having more than one sequence, constraint and conversion clause per transform, wondering if we need to specify the arguments `iop1`, `seq1`, `cons1` in the agument lists for every transform. Instead, is it okay to make them implicit?

Reply:
Yes, it's a question of language design/style. Within the constraints you mention there is no technical need for the arguments.

Reply:
For now I'm going to leave the arguments as required. I need to see a parser handling scope properly. Once I have proven I can handle scope for large/complex translation units I'll revisit this. It is marked as part of the TODO section.